File: backend/app/api/routes/access_tokens.py
import uuid
from typing import Any

from fastapi import APIRouter, HTTPException

from app import crud
from app.api.deps import CurrentUserJwtOnly, SessionDep
from app.models import (
    AccessTokenCreate,
    AccessTokenCreated,
    AccessTokenPublic,
    AccessTokensPublic,
    Message,
)

router = APIRouter()


@router.post("/", response_model=AccessTokenCreated)
def create_access_token(
    *,
    session: SessionDep,
    current_user: CurrentUserJwtOnly,
    token_in: AccessTokenCreate,
) -> Any:
    """
    Create a new access token.
    """
    db_token, token = crud.create_access_token(
        session=session, token_create=token_in, owner_id=current_user.id
    )
    return AccessTokenCreated(
        token=token, access_token=AccessTokenPublic.model_validate(db_token)
    )


@router.get("/", response_model=AccessTokensPublic)
def read_access_tokens(
    session: SessionDep,
    current_user: CurrentUserJwtOnly,
    skip: int = 0,
    limit: int = 100,
) -> Any:
    """
    Retrieve access tokens for the current user.
    """
    tokens = crud.get_access_tokens_by_user(
        session=session, user_id=current_user.id, skip=skip, limit=limit
    )
    return AccessTokensPublic(data=tokens, count=len(tokens))


@router.get("/{token_id}", response_model=AccessTokenPublic)
def read_access_token(
    *, session: SessionDep, current_user: CurrentUserJwtOnly, token_id: uuid.UUID
) -> Any:
    """
    Get access token by ID.
    """
    token = crud.get_access_token_by_id(
        session=session, token_id=token_id, user_id=current_user.id
    )
    if not token:
        raise HTTPException(status_code=404, detail="Access token not found")
    return token


@router.delete("/{token_id}", response_model=Message)
def revoke_access_token(
    *, session: SessionDep, current_user: CurrentUserJwtOnly, token_id: uuid.UUID
) -> Any:
    """
    Revoke an access token.
    """
    success = crud.revoke_access_token(
        session=session, token_id=token_id, user_id=current_user.id
    )
    if not success:
        raise HTTPException(status_code=404, detail="Access token not found")
    return Message(message="Access token revoked successfully")

File: frontend/src/services/accessTokens.ts
import type { CancelablePromise } from "@/client/core/CancelablePromise"
import { OpenAPI } from "@/client/core/OpenAPI"
import { request as __request } from "@/client/core/request"

export interface AccessTokenCreate {
  name: string
  expires_at: string
}

export interface AccessTokenPublic {
  id: string
  name: string
  prefix: string
  is_active: boolean
  expires_at: string
  last_used_at: string | null
  created_at: string
}

export interface AccessTokenCreated {
  token: string
  access_token: AccessTokenPublic
}

export interface AccessTokensPublic {
  data: AccessTokenPublic[]
  count: number
}

export interface Message {
  message: string
}

/**
 * Create Access Token
 * Create a new access token.
 */
export function createAccessToken(data: {
  requestBody: AccessTokenCreate
}): CancelablePromise<AccessTokenCreated> {
  return __request(OpenAPI, {
    method: "POST",
    url: "/api/v1/access-tokens/",
    body: data.requestBody,
    mediaType: "application/json",
    errors: {
      422: "Validation Error",
    },
  })
}

/**
 * Read Access Tokens
 * Retrieve access tokens for the current user.
 */
export function readAccessTokens(
  data: { skip?: number; limit?: number } = {},
): CancelablePromise<AccessTokensPublic> {
  return __request(OpenAPI, {
    method: "GET",
    url: "/api/v1/access-tokens/",
    query: {
      skip: data.skip,
      limit: data.limit,
    },
    errors: {
      422: "Validation Error",
    },
  })
}

/**
 * Read Access Token
 * Get access token by ID.
 */
export function readAccessToken(data: {
  tokenId: string
}): CancelablePromise<AccessTokenPublic> {
  return __request(OpenAPI, {
    method: "GET",
    url: "/api/v1/access-tokens/{token_id}",
    path: {
      token_id: data.tokenId,
    },
    errors: {
      422: "Validation Error",
    },
  })
}

/**
 * Revoke Access Token
 * Revoke an access token.
 */
export function revokeAccessToken(data: {
  tokenId: string
}): CancelablePromise<Message> {
  return __request(OpenAPI, {
    method: "DELETE",
    url: "/api/v1/access-tokens/{token_id}",
    path: {
      token_id: data.tokenId,
    },
    errors: {
      422: "Validation Error",
    },
  })
}
